### PR TITLE
REFACTOR: General component overhaul

### PR DESCRIPTION
**Pull request by pmusaraj**
- Uses `topic-navigation` and `after-topic-progress` outlets for better placement of TOC and more stability
  - expanded/collapsed state of component relies on core's full vs minimized timeline placement, so it should be more stable
- Moves JS code into separate files
- Simplifies/centralizes SCSS (and avoids
- Removes all usage of jQuery

### CODE REVIEW

File: javascripts/discourse/initializers/disco-toc-composer.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { withPluginApi } from "discourse/lib/plugin-api";
+
+export default {
+  name: "disco-toc-composer",
+
+  initialize() {
+    withPluginApi("1.0.0", (api) => {
+      const currentUser = api.getCurrentUser();
+      const currentUserTL = currentUser ? currentUser.trust_level : "";
```
Comment (eviltrout) [inline javascripts/discourse/initializers/disco-toc-composer.js:9]: It's a bit weird to default no user to an empty string. Maybe the code should `return` if there is no `currentUser`

**Review by eviltrout — APPROVED** [review]

File: javascripts/discourse/initializers/disco-toc-composer.js (hunk 1 above)
Comment (pmusaraj) [reply javascripts/discourse/initializers/disco-toc-composer.js]: Nice catch, no need to do anything here for anons, indeed. Fixed!

File: javascripts/discourse/initializers/disco-toc-main.js [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+      api.onAppEvent("topic:current-post-scrolled", (args) => {
+        if (args.postIndex !== 1) {
+          return;
+        }
+
+        const headings = document.querySelectorAll(".d-toc-post-heading");
+        let closestHeadingDistance = null;
+        let closestHeading = null;
+
+        headings.forEach(function (heading) {
```
Comment (CvX) [inline javascripts/discourse/initializers/disco-toc-main.js:100]: Arrow function could be used here

File: test/acceptance/toc-composer-test.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+  test("Can use TOC when creating a topic", async function (assert) {
+    await visit("/");
+    await click("#create-topic");
+    const toolbarPopupMenu = selectKit(".toolbar-popup-menu-options");
+    await toolbarPopupMenu.expand();
+    await toolbarPopupMenu.selectRowByValue("insertDtoc");
+
+    assert.ok(
+      query(".d-editor-input").value.indexOf('data-theme-toc="true"') >= 0
```
Comment (CvX) [inline test/acceptance/toc-composer-test.js:17]:
```suggestion
      query(".d-editor-input").value.includes('data-theme-toc="true"')
```

File: package.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "DiscoTOC",
+  "version": "2.0.0",
+  "repository": "https://github.com/discourse/DiscoTOC",
+  "author": "Discourse",
+  "license": "MIT",
+  "devDependencies": {
+    "eslint-config-discourse": "^1.1.8"
```
Comment (CvX) [inline package.json:8]:
```suggestion
    "eslint-config-discourse": "^1.1.9"
```

File: javascripts/discourse/initializers/disco-toc-main.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            closestHeadingDistance = distance;
+            closestHeading = heading;
+          } else {
+            return false;
+          }
+        });
+
+        if (closestHeading) {
+          const listItems = document.querySelectorAll(`#d-toc li`);
+          for (let i = 0; i < listItems.length; i++) {
```
Comment (CvX) [inline javascripts/discourse/initializers/disco-toc-main.js:117]: This and the other loop should probably be replaced with `for-of` or `forEach`

**Review by CvX — APPROVED** [review]

File: javascripts/discourse/connectors/after-topic-progress/d-toc-mini.hbs [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<a {{action "showTOCOverlay"}} href class="btn btn-primary">
```
Comment (jjaffeux) [inline javascripts/discourse/connectors/after-topic-progress/d-toc-mini.hbs:1]: couldn't we use a d-button here directly ?

**Pull request merged**